README.md
# ❆ faf-ice-adapter ❆

 A P2P connection proxy for Supreme Commander: Forged Alliance using [ICE](https://en.wikipedia.org/wiki/Interactive_Connectivity_Establishment).

## JSONRPC Protocol
The `faf-ice-adapter` is controlled using a bi-directional [JSON-RPC](http://www.jsonrpc.org/specification) interface over TCP.

### Methods (client ➠ faf-ice-adapter)

| Name | Parameters | Returns | Description |
| --- | --- | --- | --- |
| quit | | | Gracefully shuts down the `faf-ice-adapter`. |
| hostGame | mapName (string) | | Tell the game to create the lobby and host game on Lobby-State. |
| joinGame | remotePlayerLogin (string), remotePlayerId (int) | | Tell the game to create the Lobby, create a PeerRelay in answer mode and join the remote game. |
| connectToPeer | remotePlayerLogin (string), remotePlayerId (int), offer (bool)| | Create a PeerRelay and tell the game to connect to the remote peer with offer/answer mode. |
| disconnectFromPeer | remotePlayerId (int)| | Destroy PeerRelay and tell the game to disconnect from the remote peer. |
| setLobbyInitMode | lobbyInitMode (string): "normal" or "auto" | | Set the lobby mode the game will use. Supported values are "normal" for normal lobby and "auto" for automatch lobby (aka ladder). |
| iceMsg | remotePlayerId (int), msg (object) | | Add the remote ICE message to the PeerRelay to establish a connection. |
| sendToGpgNet | header (string), chunks (array) | | Send an arbitrary message to the game. |
| setIceServers | iceServers (array) | | ICE server array for use in webrtc. Must be called before joinGame/connectToPeer. See https://developer.mozilla.org/en-US/docs/Web/API/RTCIceServer |
| status | | [status structure](#status-structure) | Polls the current status of the `faf-ice-adapter`. |

### Notifications (faf-ice-adapter ➠ client )
| Name | Parameters | Description |
| --- | --- | --- |
| onConnectionStateChanged | "Connected"/"Disconnected" (string) | The game connected to the internal GPGNetServer. |
| onGpgNetMessageReceived | header (string), chunks (array) | The game sent a message to the `faf-ice-adapter` via the internal GPGNetServer. |
| onIceMsg | localPlayerId (int), remotePlayerId (int), msg (object) | The PeerRelays gathered a local ICE message for connecting to the remote player. This message must be forwarded to the remote peer and set using the `iceMsg` command. |
| onIceConnectionStateChanged | localPlayerId (int), remotePlayerId (int), state (string) | See https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/iceConnectionState |
| onConnected | localPlayerId (int), remotePlayerId (int), connected (bool) | Informs the client that ICE connectivity to the peer is established or unestablished. |

#### Status structure
```
{
"version" : /* string: faf-ice-adapter version */
"ice_servers" : /* the ICE servers set using `setIceServers` */
"lobby_port" : /* the actual game lobby UDP port. Should match --lobby-port option if non-zero port is specified. */
"init_mode" : /* the current init mode. See setLobbyInitMode */
"options" : /* The specified commandline options */
"gpgnet" : { /* The GPGNet state */
  "local_port" : /* int: The port the game should connect to via /gpgnet 127.0.0.1:port */
  "connected" : /* boolean: Is the game connected? */
  "game_state" : /* string: The last received "GameState" */
  "task_string" : /* string: A string describing the task/role of the game (joining/hosting)*/
  }
"relays" : [/* An array of relay information for each peer */
  {
    "remote_player_id" : /* int: The ID of the remote player */
    "remote_player_login" : /* string: The name of the remote player */
    "local_game_udp_port" : /* int: The UDP port opened for the game to connect to */
    "ice": {/* ICE state information for this peer */
      "offerer": /* bool: one peer is always offerer, one answerer */
      "state": /* string: The connection state https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/iceConnectionState */
      "gathering_state": /* string: The gathering state https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/iceGatheringState */
      "datachannel_state": /* string: The state of the https://developer.mozilla.org/en-US/docs/Web/API/RTCDataChannel */
      "connected": /* bool: Is the peer connected? Needs to be in sync with the remote peer. */
      "loc_cand_addr": /* string: The local address used for the connection */
      "rem_cand_addr": /* string: The remote address used for the connection */
      "loc_cand_type": /* string: The type of the local candidate 'local'/'stun'/'relay' */
      "rem_cand_type": /* string: The type of the remote candidate 'local'/'stun'/'relay' */
      "time_to_connected": /* double: The time it took to connect to the peer in seconds */
      }
    },
  ...
  ]
}
```

## Commandline invocation
The first two commandline arguments `--id` and `--login` must be specified like this: `faf-ice-adapter -i 3 -l "Rhiza"`
The full commandline help text is:
```
faf-ice-adapter usage:
--help                               produce help message
--id arg                             set the ID of the local player
--login arg                          set the login of the local player, e.g. "Rhiza"
--rpc-port arg (=7236)               set the port of internal JSON-RPC server
--gpgnet-port arg (=0)               set the port of internal GPGNet server
--lobby-port arg (=0)                set the port the game lobby should use for incoming UDP packets from the PeerRelay
--log-directory arg                  set a log directory to write ice_adapter_0 log files
```

## Example usage sequence

| Step | Player 1 "Alice" | Player 2 "Bob" |
| --- | --- | --- |
| 1 | Start the client |  |
| 2 | The client starts `faf-ice-adapter` and connects to the JSONRPC server |  |
| 3 | The client starts the game and makes it connect to the GPGNet server of the `faf-ice-adapter` using `/gpgnet 127.0.0.1:7237` commandline argument for `ForgedAlliance.exe` |  |
| 3.1 | The game sends `GameState Idle` | |
| 3.2 | The client sends `CreateLobby ...` | |
| 3.3 | The game sends `GameState Lobby` | |
| 4 | The client sends `hostGame('monument_valley.v0001')`||
| 5 | The game should now wait in the lobby and the client receives a lot of `onGpgNetMessageReceived` notifications from `faf-ice-adapter`||
| 6 |  | Now Bob want to join Alices game, starts the client, the client starts `faf-ice-adapter` and the game like Alice did. |
| 6.1 | | The game sends `GameState Idle` |
| 6.2 | | The client sends `CreateLobby ...` |
| 6.3 | | The game sends `GameState Lobby` |
| 7 | The client sends `connectToPeer('Bob', 2, true)` | The client sends `joinGame('Alice', 1)` |
| 8 | The game should connect to the internal PeerRelay, and shows that it's connecting to the peer. | The game should connect to the internal PeerRelay, and shows that it's connecting to the peer. |
| 9 | The client receives multiple `onIceMsg(1, 2, someIceMsg)`notifications and must now transfer these messages string ordered to the peer. | The client receives multiple `onIceMsg(1, 2, someIceMsg)`notifications and must now transfer these messages string ordered to the peer. |
| 10 | The client must set the transferred ICE messages for the peer using `iceMsg(2, someIceMsg)`. | The client must set the transferred ICE messages for the peer using `iceMsg(1, someIceMsg)`. |
| 11 | The client received multiple `iceConnectionStateChanged(...)` notifications which would finally show the `'Connected'` or `'Complete'` state, which should also let the game connect to the peer. Another indicator for a connection is the `onDatachannelOpen` notification.| The client received multiple `iceConnectionStateChanged(...)` notifications which would finally show the `'Connected'` or `'Complete'` state, which should also let the game connect to the peer. |

## Building `faf-ice-adapter`
###  Linux
Webrtc is build using clang and linked against clangs libc++, so you need to use these for building the ice-adapter.
1. Build [libwebrtc](https://github.com/FAForever/libwebrtc) with `cmake -G Ninja -DCMAKE_BUILD_TYPE=Release -DWEBRTC_BRANCH_HEAD=refs/branch-heads/62 -DBUILD_TESTS=ON`
2. Build the ice-adapter with `cmake -DCMAKE_CXX_COMPILER="/usr/bin/clang++" -DCMAKE_CXX_FLAGS="-stdlib=libc++" -DWEBRTC_INCLUDE_DIRS="path/to/webrtc/include" -DWEBRTC_LIBRARIES="path/to/webrtc/lib/libwebrtc.a" -DCMAKE_BUILD_TYPE=Release` and `make`
###  Windows
1. Download and extract [latest libwebrtc win32 release zip file](https://github.com/FAForever/libwebrtc/releases/latest).
2. Install Visual Studio 2015 compilers and open x86 shell.
3. Build the ice-adapter using `cmake -DWEBRTC_INCLUDE_DIRS="path/to/webrtc/include" -DWEBRTC_LIBRARIES="path/to/webrtc/lib/libwebrtc.lib" -DCMAKE_BUILD_TYPE=Release`


CHANGELOG.md
v6.3.4
======
- include "ice_servers" in `status()` struct

v6.3.3
======
- attempt to fix crash on _reinitPeerconnection
- logging and console output fixes

v6.3.2
======

- update to webrtc 68
- connectivitychecker ping now has some startup time

v6.3.1
======

- fix messed up TURN connections

v6.3.0
======

- reconnect is now back to recreating the whole PeerConnection instead
  of using ICE restart to fix memory leak issues

v6.2.6
======

- fix stats not always collected
- refactor connectivitychecker
- print version on commandline

v6.2.5
======

- enable travis CI build for linux binary

v6.2.4
======

- build against current libwebrtc

v6.2.3
======

- adjust for webrtc 66
- improve GPGNetMessage parsing

v6.2.2
======

- use a custom ping

v6.2.1
======

- more work on reconnect
- more status information

v6.2.0
======

- change reconnect behaviour

v6.1.3
======

- fixing the GPGNetServer
- a lot more testing

v6.1.2
======

- fix not using ICE servers

v6.1.1
======

- testing and debugging

v6.1.0
======

- fix accidentially running multiple threads
- fix automatic lobby port
- create new faf-ice-testclient executable to allow unattended testing

v6.0.0
======

- rewrite as pure C++ implementation only depending on a native WebRTC build

v5.6.1
======

- use updates webrtc

v5.6.0
======

- remove Testclient and simplify CI build

v5.5.3
======

- refactor logging in PeerRelay

v5.5.2
======

- #44 don't interrupt 'checking' state after 10 second timeout
- #43 fix logfile size

v5.5.1
======

- do more logging

v5.5.0
======

- add `setInitMode` API method to enable auto-lobby for ladder/matchmaking

v5.4.2
======

- use updated node-webrtc binaries

v5.4.1
======

- rework closing RTCPeerConnections to fix #34

v5.4.0
======

- create `setIceServers` API method, which must be called before connecting to peers
- work on reconnect

v5.3.1
======

- create better platform archives

v5.3.0
======

- initial release of automatic reconnect in testclient, testserver and ice-adapter
- create this CHANGELOG

v5.2.0
======

- use hyphens instead of underscore in options
- fix build of `faf-ice-testclient` on travis CI

v5.1.4
======

- make the established peer connection type accessible
- add creation of dummy peerrelay at ice-adapter startup to trigger windows firewall prompt

v5.1.3
======

- fix long string parsing buffer error in GPGNetMessage which made subsequent games fail
- reset gpgnetstate

v5.1.2
======

- reset gameTaskString on disconnect

v5.1.1
======

- use new x86 webrtc node module for testclient release archive on travis CI

v5.1.0
======

- create a JS bundle of `faf-ice-adapter` on travis CI
- create a testclient release archive on travis CI

v5.0.2
======

- disconnect peers on game exit
- fix datachannel_open status

v5.0.1
======

- make JSON parsing more compatible
- cleanup

v5.0.0
======

- first nodejs-based release

v4.0.0
======

- refactor to use one stream per peer and a single agent

v3.0.0
======

- created `faf-ice-testserver` and `faf-ice-testclient`

v2.0.0
======

- include uPnP and port range options


v1.0.0
======

- first libnice based release


options.ts
import * as optimist from 'optimist';

export let options = optimist
  .usage('Usage: $0 --id [num] --login [string]')
  .demand(['id', 'login'])
  .describe('id', 'set the ID of the local player')
  .describe('login', 'set the login of the local player, e.g. "Rhiza"')
  .default('rpc_port', 7236)
  .describe('rpc_port', 'set the port of internal JSON-RPC server')
  .default('gpgnet_port', 7237)
  .describe('gpgnet_port', 'set the port of internal GPGNet server')
  .default('lobby_port', 7238)
  .describe('lobby_port', 'set the port the game lobby should use for incoming UDP packets from the PeerRelay')
  .default('stun_server', 'test.faforever.com')
  .describe('stun_server', 'set the STUN server hostname')
  .default('turn_server', 'test.faforever.com')
  .describe('turn_server', 'set the TURN server hostname')
  .default('turn_user', 'test')
  .describe('turn_user', 'set the TURN username')
  .default('turn_pass', 'test')
  .describe('turn_pass', 'set the TURN password')
  .default('log_file', '')
  .describe('log_file', 'set a verbose log file')
  .argv;

export default options;

